dns/client.go

```diff
@@ -102,13 +148,141 @@ func (r *Client) Resolve(
 	return ips, nil
 }
 
-// Lookup resolves a domain name to a slice of IP addresses using the system's
-// default resolver.
+// Lookup resolves a domain name to a slice of IP addresses.
+//
+// It prefers the k6 VU dialer's resolver if available, so that k6 networking
+// options (e.g. blockHostnames) are honored. It falls back to the system's
+// default resolver otherwise.
 func (r *Client) Lookup(ctx context.Context, hostname string) ([]string, error) {
-	ips, err := net.DefaultResolver.LookupHost(ctx, hostname)
+	vustate := r.vu.State()
+	if vustate == nil {
+		return nil, common.NewInitContextError("using lookup in the init context is not supported")
+	}
+
+	// If k6's dialer is available, honor its blocked hostnames configuration.
+	if d, ok := vustate.Dialer.(*netext.Dialer); ok && d != nil && d.BlockedHostnames != nil {
+		normalized := strings.TrimSuffix(strings.ToLower(hostname), ".")
+		if _, blocked := d.BlockedHostnames.Contains(normalized); blocked {
+			return nil, &Error{
+				Name:    "BlockedHostname",
+				Message: fmt.Sprintf("blocked hostname: %s", normalized),
+				Kind:    Refused,
+			}
+		}
+	}
+
+	if vustate.Dialer == nil {
```

Comment:
Nit; arguably this could be combined with the condition statement above, or moved before, as I think it overlaps.

Unless I'm wrong, the `d != nil` above is the opposite of `vustate.Dialer == nil`.

Reply:
Maybe, if you do the refactor I suggested above, then you probably want to leave it as-is. Either way!

Reply:
That's a good question, looking into it 👀